programs/stake/src/stake_state.rs

```diff
@@ -407,37 +407,45 @@ impl Stake {
     ///   for credits_observed were the points paid
     pub fn calculate_points_and_credits(
         &self,
-        vote_state: &VoteState,
+        new_vote_state: &VoteState,
         stake_history: Option<&StakeHistory>,
     ) -> (u128, u64) {
-        if self.credits_observed >= vote_state.credits() {
+        // if there is no newer credits since observed, return no point
+        if new_vote_state.credits() <= self.credits_observed {
             return (0, 0);
         }
 
-        let mut credits_observed = self.credits_observed;
-        let mut points = 0u128;
-        for (epoch, credits, prev_credits) in vote_state.epoch_credits() {
+        let mut points = 0;
+        let mut new_credits_observed = self.credits_observed;
+
+        for (epoch, final_epoch_credits, initial_epoch_credits) in
```

Comment:
the naming of `credits` and `prev_credits` here was a bit unhelpful in this context, imo.

what's important here is that they're together forming an interval because we're calculating overlaps between it and `self.credit_observed`.

So, I prefixed both of them with prefixes. (they used symmetrically in the logic but the one has a prefix and the other not; that's another point of confution, imo). `final_` and `initial_` is chosen over `last_/first_` and `end_/start_`:

- these epoch_credit interval's bounds are rather more rigid, not arbitrary, like they generally must not overlap with each other and properly ordered, etc. For example, this is not like lockout_intervals. So let's use somewhat formal-sounding wording here.
- it's less odd to see (final, initial) order than `(end, start)` because the `final` wording has more semantic significance than `initial`; usually `end` and `start` have equal balance. Otherwise, we have to change ABI to make `first, last` to happen...